specs/ChainSpec/LightClient.md

```diff
@@ -0,0 +1,170 @@
+    pub approvals_after_next: Vec<Option<Signature>>,
+}
+```
+
+Recall that the hash of the block is
+
+```rust
+sha256(concat(
+    prev_hash,
+    sha256(concat(
+        sha256(borsh(inner_lite)),
+        sha256(borsh(inner_rest))
+    ))
+))
+```
+
+The fields `prev_block_hash`, `next_block_inner_hash` and `inner_rest_hash` are used to reconstruct the hashes of the current and next block, and the approvals that will be signed, in the following way (where `block_view` is an instance of `LightClientBlockView`):
+
+```python
+def reconstruct_light_client_block_view_fields(block_view):
+    current_block_hash = sha256(concat(
+        block_view.prev_block_hash,
+        sha256(concat(
+            sha256(borsh(block_view.inner_lite)),
+            block_view.inner_rest_hash,
+        ))
+    ))
+
+    next_block_hash = sha256(concat(
+        current_block_hash,
+        block_view.next_block_inner_hash
+    ))
+
+    approval_message = concat(
+        borsh(ApprovalInner::Endorsement(next_block_hash)),
+        little_endian(block_view.inner_lite.height + 2)
+    )
+
+    return (current_block_hash, next_block_hash, approval_message)
+```
+
+The light client updates its head with the information from `LightClientBlockView` iff:
+
+1. The height of the block is higher than the height of the current head;
+2. The epoch of the block is equal to the `epoch_id` or `next_epoch_id` known for the current head;
+3. If the epoch of the block is equal to the `next_epoch_id` of the head, then `next_bps` is not `None`;
+4. `approvals_after_next` contain valid signatures on `approval_message` from the block producers of the corresponding epoch (see next section);
+5. The signatures present in `approvals_after_next` correspond to more than 2/3 of the total stake (see next section).
+6. If `next_bps` is not none, `sha256(borsh(next_bps))` corresponds to the `next_bp_hash` in `inner_lite`.
+
+```python
+def validate_and_update_head(block_view):
+    global head
+    global epoch_block_producers_map
+
+    current_block_hash, next_block_hash, approval_message = reconstruct_light_client_block_view_fields
+
+    # (1)
+    if block_view.inner_lite.height <= head.inner_lite.height:
+        return False
+
+    # (2)
+    if block_view.inner_lite.epoch_id not in [head.inner_lite.epoch_id, head.inner_lite.next_epoch_id]:
+        return False
+
+    # (3)
+    if block_view.inner_lite.epoch_id == head.inner_lite.next_epoch_id and block_view.next_bps is None:
+        return False
+
+    # (4) and (5)
+    total_stake = 0
+    approved_stake = 0
+
+    epoch_block_producers = epoch_block_producers_map[block_view.inner_lite.epoch_id]
+    for maybe_signature, block_producer in zip(block_view.approvals_after_next, epoch_block_producers):
+        total_stake += block_producer.stake
+
+        if maybe_signature is None:
+            continue
+
+        approved_stake += block_producer.stake
+        if not verify_signature(
```

Comment:
We should check whether it was a skip message here and not return False if it was.

Reply:
Not sure I understand.
All the approval messages in the block are identical, and for the purposes of a light client block they are all endorsements of a very particular form, exactly the form that we validate the signature against here.

Reply:
I don't remember what `ApprovalInner::Skip` corresponds to but it looks like this is something a validator can sign instead of `ApprovalInner:: Endorsement` and then `approvals_after_next` might contain `None`, and signatures of `Endorsement` and `Skip`. We are only taking care of `None` and signature of `Endorsement` here.

Reply:
Incidentally, this is one of the few components of the chain that have a full detailed spec:

https://nomicon.io/ChainSpec/Consensus.html

The block producer might have signed endorsements or skips, but a paritcular **block** can only include signatures on a very particular kind of approval (endorsement if its previous block has sequential height, or skip if there are skipped heights). A final block always has two blocks ahead of it with sequential heights, and thus they always only contain signatures on endorsements.